invenio_records_files/api.py

```diff
@@ -303,7 +283,96 @@ def files(self, data):
 
 
 class Record(_Record, FilesMixin):
-    """Define API for files manipulation using ``FilesMixin``."""
+    """Record class with associated bucket.
+
+    The record class implements a one-to-one relationship between a bucket and
+    a record. A bucket is automatically created and associated with the record
+    when the record is created with :py:data:`Record.create()` (unless
+    ``with_bucket`` is set to ``False``).
+
+    The bucket id is stored in the record metadata (by default in the
+    ``_bucket`` key). You can implement your dump/load behavior for the storing
+    the bucket id in the record (or possibly somewhere else). You do this by
+    creating a subclass of this class, and override the two classmethods
+    :py:data:`Record.dump_bucket()` and :py:data:`Record.load_bucket()`.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """Initialize the record."""
+        self._bucket = None
+        super(Record, self).__init__(*args, **kwargs)
+
+    @classmethod
+    def create(cls, data, id_=None, with_bucket=True, **kwargs):
+        """Create a record and the associated bucket.
+
+        :param with_bucket: Create a bucket automatically on record creation.
+        """
+        # Create bucket and store in record metadata.
+        if with_bucket:
+            bucket = cls.create_bucket(data)
+            if bucket:
+                cls.dump_bucket(data, bucket)
+        # Create the record
+        record = super(Record, cls).create(data, id_=id_, **kwargs)
+        # Create link between record and bucket
+        if with_bucket and bucket:
+            RecordsBuckets.create(record=record.model, bucket=bucket)
+            record._bucket = bucket
+        return record
+
+    @classmethod
+    def create_bucket(cls, data):
+        """Create a bucket for this record.
+
+        Override this method to provide more advanced bucket creation
+        capabilities. This method may return a new or existing bucket, or may
+        return None, in case no bucket should be created.
+        """
+        return Bucket.create()
+
+    @classmethod
+    def dump_bucket(cls, data, bucket):
+        """Dump the bucket id into the record metadata.
+
+        Override this method to provide custom behavior for storing the bucket
+        id in the record metadata. By default the bucket id is stored in the
+        ``_bucket`` key. If you override this method, make sure you also
+        override :py:data:`Record.load_bucket()`.
+
+        This method is called after the bucket is created, but before the
+        record is created in the database.
+
+        :param data: A dictionary of the record metadata.
+        :param bucket: The created bucket for the record.
+        """
+        data['_bucket'] = str(bucket.id)
+
+    @classmethod
+    def load_bucket(cls, record):
```

Comment:
Could `data` actually also be an instance of a `Record`? There is a symmetry between load_bucket and dump_bucket that is not quite complete. Also throughout the code base there seems to be a desire to manipulate the same dict interface when dealing with records and record-dict.

Reply:
It's tricky. ``dump_bucket()`` is called on the dict prior to the record instance being created (in order to prevent an extra sql query). ``load_bucket()`` is always called on the record itself.

Reply:
Yes, since we have these transitory states it is tricky. More thought needed...